### PR TITLE
DBZ-2637 Enhance Postgres validation checks

### PR DESCRIPTION
**Pull request by rk3rn3r**
DBZ-2637 Enhance Postgres validation checks
+ add wal_level config check
+ add LOGIN and REPLICATION roles check
+ add replication slot is already in use check

closes https://issues.redhat.com/browse/DBZ-2637

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 1]
```diff
@@ -561,7 +561,13 @@ public static SchemaRefreshMode parse(String value) {
...
             .withImportance(Importance.MEDIUM)
             .withDescription("The name of the Postgres logical decoding plugin installed on the server. " +
-                    "Supported values are '" + LogicalDecoder.DECODERBUFS.getValue() + "' and '" + LogicalDecoder.WAL2JSON.getValue() + "'. " +
+                    "Supported values are '" + LogicalDecoder.DECODERBUFS.getValue()
+                    + "', '" + LogicalDecoder.WAL2JSON.getValue()
+                    + "', '" + LogicalDecoder.PGOUTPUT.getValue()
+                    + "', '" + LogicalDecoder.WAL2JSON_STREAMING.getValue()
+                    + "', '" + LogicalDecoder.WAL2JSON_RDS.getValue()
+                    + "' and '" + LogicalDecoder.WAL2JSON_RDS_STREAMING.getValue()
+                    + "'. " +
```
Comment (rk3rn3r) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:570]: Unrelated cleanup in separate commit. Is it okay to show all possible values?

Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:570]: This is fine. I wonder if you could create a utility method that would format the allow values for any enum.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java [hunk 2]
```diff
@@ -321,15 +321,15 @@ public void initConnection() throws SQLException, InterruptedException {
...
+        // note that some of these options are only supported in Postgres 9.4+, additionally
+        // the options are not yet exported by the jdbc api wrapper, therefore, we just do
+        // this ourselves but eventually this should be moved back to the jdbc API
         // see https://github.com/pgjdbc/pgjdbc/issues/1305
 
         LOGGER.debug("Creating new replication slot '{}' for plugin '{}'", slotName, plugin);
         String tempPart = "";
-        // Exported snapshots are supported in PostgreSQL 9.4+
-        Boolean canExportSnapshot = pgConnection().haveMinimumServerVersion(ServerVersion.v9_4);
+        // Exported snapshots are supported in Postgres 9.4+
```
Comment (jpechane) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:331]: The database is really named PostgreSQL :-)

Comment (rk3rn3r) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:331]: I know. A PostgreSQL committer was my professor at university. Still, we tend to use Postgres in most places.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java (hunk 1 above)
Comment (rk3rn3r) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:570]: I can, but after holidays.

**Comment by jpechane** [comment]
@rk3rn3r Applied, thanks!

**Comment by rafalwytrykus** [comment]
Hi, I've noticed that this change seems to break Debezium on AWS RDS where some roles have non-standard names e.g. `rds_replication` instead of `replication`

**Comment by jpechane** [comment]
@rafalwytrykus Could you please raise a Jira issue?

**Comment by rk3rn3r** [comment]
thank you @rafalwytrykus! I will create the Jira issue for it

**Comment by rk3rn3r** [comment]
@rafalwytrykus I checked the query on Amazon RDS Postgres and it looks like it's working. Can you maybe review my comment in the task and validate it's the same for you? https://issues.redhat.com/browse/DBZ-2800

**Comment by rafalwytrykus** [comment]
@rk3rn3r Thank you for checking it. I am not sure about classic RDS but for Aurora the master user does not have `replication` role:
```
postgres=> SELECT rolcanlogin, rolreplication FROM pg_roles WHERE rolname = current_user;
 rolcanlogin | rolreplication
-------------+----------------
 t           | f
(1 row)
```
master is however a member of rds_replication:
```
postgres=> \du
                                                                  List of roles
       Role name       |                         Attributes                         |                          Member of
-----------------------+------------------------------------------------------------+-------------------------------------------------------------
 full_sch_udm_data     | Cannot login                                               | {}
 full_sch_udm_data_pii | Cannot login                                               | {}
 master                | Create role, Create DB                                    +| {rds_superuser,rds_replication}
                       | Password valid until infinity                              |
 rds_ad                | Cannot login                                               | {}
 rds_iam               | Cannot login                                               | {}
 rds_password          | Cannot login                                               | {}
 rds_replication       | Cannot login                                               | {}
```
The Postgres version I am using is 11.6

**Comment by RemingtonReackhof** [comment]
@rafalwytrykus @rk3rn3r I have ran into the same issue on classic RDS, running PostgreSQL 12.4-R1:
```
SELECT rolcanlogin, rolreplication FROM pg_roles WHERE rolname = current_user;
 rolcanlogin | rolreplication
-------------+----------------
 t           | f
(1 row)
```

and the master is a member of `rds_replication`:
```
\du postgres
                                List of roles
 Role name |          Attributes           |            Member of
-----------+-------------------------------+---------------------------------
 postgres  | Create role, Create DB       +| {rds_superuser,rds_replication}
           | Password valid until infinity |
```
Oddly enough, it looks like `rds_replication` itself doesn't have `rolreplication=true`
```
SELECT rolcanlogin, rolreplication FROM pg_roles WHERE rolname='rds_replication';
 rolcanlogin | rolreplication
-------------+----------------
 f           | f
(1 row)
```

**Comment by RemingtonReackhof** [comment]
I've downgraded to 1.3.1 and can confirm that the connector works given my current setup described above ^

**Comment by rk3rn3r** [comment]
Thank you @rafalwytrykus @RemingtonReackhof for checking.
Can you share the output of `SELECT * FROM pg_roles WHERE rolname = current_user;` for the master user.

When I checked on a fresh AWS RDS Postgres I got (for master user):
```
postgres=> SELECT rolcanlogin, rolreplication FROM pg_roles WHERE rolname = current_user;

 rolcanlogin | rolreplication  
-------------+----------------

 t           | t

```

**Comment by rk3rn3r** [comment]
Happy new year @rafalwytrykus @RemingtonReackhof !
And sorry to bother you again again.

Can you share the output of `SELECT * FROM pg_roles WHERE rolname = current_user;` for the master user so I can see the complete list of available roles for the current user.
Thx again!

**Comment by gunnarmorling** [comment]
See DBZ-2800 for what we'll do for 1.4.0.Final: add the RDS specific role name to the check, and also make it more lenient (i.e. log the error about missing roles, but don't make it an actual validation error).

**Comment by rafalwytrykus** [comment]
Thank you @rk3rn3r and Happy new year!

This is the output from Aurora 11.6:
```
=> SELECT * FROM pg_roles WHERE rolname = current_user;
 rolname | rolsuper | rolinherit | rolcreaterole | rolcreatedb | rolcanlogin | rolreplication | rolconnlimit | rolpassword | rolvaliduntil | rolbypassrls | rolconfig |  oid
---------+----------+------------+---------------+-------------+-------------+----------------+--------------+-------------+---------------+--------------+-----------+-------
 master  | f        | t          | t             | t           | t           | f              |           -1 | ********    | infinity      | f            |           | 16399
(1 row)
```

**Comment by rk3rn3r** [comment]
Thank you @rafalwytrykus !

I took a closer look at a AWS RDS Postgres installations today. When I understand it correctly, the debezium connector user needs to have one of these roles assigned: `rds_superuser`, `rdsadmin` or `rdsrepladmin` in order to have the rolreplication rights that allow starting a logical replication. Correct?

When I get it right, that's also why we currently recommend to use the RDS master user as debezium user. Can you confirm my findings?

**Comment by rk3rn3r** [comment]
@rafalwytrykus @RemingtonReackhof Can you have a look at #2037 and also at the tasks https://issues.redhat.com/browse/DBZ-2893 and https://issues.redhat.com/browse/DBZ-2894. This should fix the issue with AWS, and also with other environments later in a more generic way.b WDYT?

**Comment by RemingtonReackhof** [comment]
@rk3rn3r apologies for the late response. #2037 looks like the right approach, I appreciate the fix 👍 . I'm having issues creating an account on redhat so I can't take a look at the issues right now. I appreciate the work on this :)
